### PR TITLE
Refine Loss in Model

### PR DESCRIPTION
**Pull request by qingqing01**
- 1  去掉 shape_hints，并增加 Input
- 2  去掉Loss种的 infer_shape and infer_dtype
- 3  去掉train/test/eval接口的device和 device_ids，放到prepare接口中
- 4  device和device_ids支持设置为None, 依据安装的Paddle版本自动选择
- 5  静态图在prepare中构建program
- 6  去掉lazy加载。
- 7  prepare接口，optimizer和loss_function支持设置为None，例如eval和test时可以没有opt和loss

Now, there is no unit tests for any module, will add tests in the PR of Fit API.

简单例子：

```
        # 对于静态图，必需声明inputs和labels。动态图也能使用
        inputs=[Input([None, 1, 28, 28], 'float32', name='image')]   # 单个输入，可不用[]
        labels=[Input([None, 1], 'int64', name='label')]             # 单个label，可不用[]
        model = MNIST()
        optim = Momentum(learning_rate=FLAGS.lr, momentum=.9,
                         parameter_list=model.parameters())
        model.prepare(optim, CrossEntropy()，inputs, labels)
```

原先实现方式:
- 优点:
   - 一些case自动推断inputs和label的shape、dtype: 
      - 简单的情况，如图像分类，固定长度训练，通过输入numpy.array的数据自动输入shape和dtype。
      - 当label和组网outputs (通常是forward的返回)shape和dtype相同，自动推断。
   - label和Model可以完全解耦。

- 缺点:
   - 无法自动推断的情况:
       - forward函数需设置装饰器shape_hints进行推断。
       - Loss需实现infer_shape和infer_dtype进行推断。
       - 两处设置地方也不统一。  
    - shape_hints无法配置化设置shape参数。
    - 使得使用DataLoader变得困难:  当DataLoader非iterable模式，无法获取到dtype。
    - 在没有拿到数据，网络运行前，静态图中，对graph的变换导致变得困难。

- train和eval复用Model导致的缺点:  train和eval复用同一个Model，但输入不同，并且输入shape无法自动推断时
  - shape_hints必需指定所有forward输入(包含train和eval)
  - train和eval的reader必需返回所有输入的数据，如果没有数据，也必须为None

当前的实现方式:
- 优点:
   - 可解决原先实现方式的缺点
- 缺点:
   - 对于静态图，简单的case，也需要声明inputs和labels。
   - 对于静态图，label作为了Model的输入，没有完全解耦。

- train和eval复用Model导致的缺点仍然存在

### CODE REVIEW

File: mnist2.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            act="softmax")
+
+    def forward(self, inputs, label):
+        x = self._simple_img_conv_pool_1(inputs)
+        x = self._simple_img_conv_pool_2(x)
+        x = fluid.layers.flatten(x, axis=1)
+        x = self._fc(x)
+        loss = fluid.layers.cross_entropy(x, label)
+        loss = fluid.layers.mean(loss)
+        self.set_loss(loss)
```
Comment (qingqing01) [inline mnist2.py]: this is only an example to show how to use `set_loss`, will not merge code.

Comment (willthefrog) [reply mnist2.py]: how to run `test` in this fashion? `label=None`

File: model.py [hunk 2]
```diff
@@ -86,24 +69,21 @@ def __call__(self, outputs, labels):
...
+            losses = [fluid.layers.reduce_sum(l) for l in losses]
+        return losses
 
 
 class CrossEntropy(Loss):
-    def __init__(self):
+    def __init__(self, average=True):
         super(CrossEntropy, self).__init__()
 
-    def infer_shape(self, outputs):
```
Comment (willthefrog) [inline model.py]: if `CrossEntropy` is wrapped anyway, why not keep the these inference methods? otherwise user need to define label shape for every model using `CrossEntropy` loss

File: model.py [hunk 3]
```diff
@@ -257,17 +234,22 @@ def _run(self, inputs, labels=None, device='CPU', device_ids=None):
...
             + " does not match number of arguments of `forward` method"
 
         if self._progs.get(self.mode, None) is None:
-            self._make_program(self._infer_input_vars(inputs))
+            if self.model._inputs is None:
+                raise ValueError("The inputs of Model must be not None.")
+            self._input_vars = [
+                k.forward() for k in to_list(self.model._inputs)
+            ]
+            self._make_program(self._input_vars)
```
Comment (willthefrog) [inline model.py]: isn't this supposed to happen in `prepare` now?

File: model.py [hunk 4]
```diff
@@ -221,18 +197,19 @@ def _load(path):
             "optimizer saved in dygraph mode is not usable in static graph"
 
         if self._executor is not None:
-           self._load_optimizer(optim_state)
+            self._load_optimizer(optim_state)
         else:
-           self._lazy_load_optimizer = optim_state
+            self._lazy_load_optimizer = optim_state
```
Comment (willthefrog) [inline model.py]: if program construction happens in `prepare`,  these lazy loading mechanism are just cruft, and should be removed for simplification.

**Comment by willthefrog** [comment]
also, as I understand,  non-iterable dataloader is supposed to be used in `fit()` (would be quite cumbersome to use it in `train` per batch fashion), in that case, we should be able to find some workarounds and make it work in the original design.

File: model.py (hunk 3 above)
Comment (qingqing01) [reply model.py]: This PR is only for loss, not change other module.  If this PR is accepted, can do this.

File: model.py (hunk 4 above)
Comment (qingqing01) [reply model.py]: As above reply.

File: mnist2.py (hunk 1 above)
Comment (qingqing01) [reply mnist2.py]: In inference of MNIST,  like YOLOv3, can use `if self.mode != 'test':` to control. 
`set_loss` is a way of use, but not the only way.

Comment (qingqing01) [reply mnist2.py]: Remove set_loss in Model

File: model.py (hunk 3 above)
Comment (qingqing01) [reply model.py]: Move to prepare

File: model.py [hunk 5]
```diff
@@ -51,33 +53,21 @@ def extract_args(func):
         return inspect.getargspec(func)[0]
 
 
-def shape_hints(**hints):
-    assert hints, "hints can not be empty"
-    assert all(isinstance(h, (list, tuple)) for h in hints.values()), \
-        "shape hint must be a list or tuple"
+class Input(fluid.dygraph.Layer):
```
Comment (willthefrog) [inline model.py]: what is the point of making it a class instead of a plain function?

File: model.py [hunk 6]
```diff
@@ -374,38 +359,36 @@ def _make_program(self, inputs):
...
-        dtypes = self.model._loss_function.infer_dtype(outputs)
-        label_vars = []
-        for idx, (shape, dtype) in enumerate(zip(shapes, dtypes)):
-            name = '__label{}'.format(idx)
-            label_vars.append(fluid.data(name, shape, dtype))
-        return label_vars
-
-    def _compile_and_initialize(self, prog, device='CPU', device_ids=None):
+    def prepare(self):
         compiled_prog = self._compiled_progs.get(self.mode, None)
```
Comment (willthefrog) [inline model.py]: only compile for current mode? what happens when use set mode before calling `prepare`?

File: model.py [hunk 7]
```diff
@@ -307,21 +288,18 @@ def _set_var(self, var, ndarray):
 
         t.set(ndarray, place)
 
-    def _run(self, inputs, labels=None, device='CPU', device_ids=None):
+    def _run(self, inputs, labels=None, device='CPU'):
+        compiled_prog = self.prepare()
```
Comment (willthefrog) [inline model.py]: I think the whole point is to build program ahead of time, why is it still running `prepare` here?

File: model.py (hunk 5 above)
Comment (qingqing01) [reply model.py]: In order to add data Variable into Program when only calling forward in static graph.  If a function to call fluid.data(), needs to clone var.

File: model.py (hunk 7 above)
Comment (qingqing01) [reply model.py]: 第一次调用prepare时，并不知道后续运行模式，是只train/eval/test，还是有模式混合。因此只，构建默认mode的Program，并缓存。后续其他mode，采用lazy模式构建，即用到时再构建。

File: model.py (hunk 6 above)
Comment (qingqing01) [reply model.py]: 如下回复。

File: model.py [hunk 8]
```diff
@@ -331,16 +309,25 @@ def _run(self, inputs, labels=None, device='CPU', device_ids=None):
...
+                                 feed=feed,
+                                 fetch_list=fetch_list)
         if self.mode == 'test':
             return out[:num_output]
         else:
             return out[:num_output], out[num_output:]
 
+    def _get_loss(self, outputs):
+        assert self.model._loss_function
+        label_vars = [k.forward() for k in to_list(self.model._labels)]
```
Comment (guoshengCS) [inline model.py]: `model._labels` may also be a dict and processed accordingly.

File: model.py [hunk 9]
```diff
@@ -374,38 +359,36 @@ def _make_program(self, inputs):
...
             return compiled_prog
 
-        places = [device.lower() == 'gpu' and fluid.CUDAPlace(i)
-                  or fluid.CPUPlace() for i in device_ids]
+        if isinstance(self.model._inputs, dict):
+            ins = [self.model._inputs[n] \
+                for n in extract_args(self.model.forward) if n != 'self']
+        else:
+            ins = self.model._inputs
+        self._input_vars[self.mode] = [k.forward() for k in to_list(ins)]
```
Comment (guoshengCS) [inline model.py]: It seems that `_input_vars` for different modes all has same contents since `model._inputs` can only  be set once in `model.prepare`.  However, it doesn't matter.

File: model.py [hunk 10]
```diff
@@ -307,21 +288,18 @@ def _set_var(self, var, ndarray):
 
         t.set(ndarray, place)
 
-    def _run(self, inputs, labels=None, device='CPU', device_ids=None):
+    def _run(self, inputs, labels=None, device='CPU'):
```
Comment (guoshengCS) [inline model.py]: The argument `device` seems unused here. Can we remove it

File: mnist.py [hunk 11]
```diff
@@ -137,13 +137,16 @@ def null_guard():
...
         model = MNIST()
-        optim = Momentum(learning_rate=FLAGS.lr, momentum=.9,
-                         parameter_list=model.parameters())
-        model.prepare(optim, CrossEntropy())
+        optim = Momentum(
+            learning_rate=FLAGS.lr,
+            momentum=.9,
+            parameter_list=model.parameters())
+        inputs = [Input([None, 1, 28, 28], 'float32', name='image')]
+        #inputs = {'inputs': Input([None, 1, 28, 28], 'float32', name='image')}
```
Comment (heavengate) [inline mnist.py]: delete

File: model.py [hunk 12]
```diff
@@ -219,19 +200,18 @@ def _load(path):
         if optim_state is None:
             return
 
-        if self._executor is not None:
-            self._load_optimizer(optim_state)
-        else:
-            self._lazy_load_optimizer = optim_state
+        assert self._executor
```
Comment (heavengate) [inline model.py]: add assert message

Comment (willthefrog) [reply model.py]: As pointed out above, clear error message should be provided when`load` is called before `prepare`.
Also "removing lazy loading" could be tricky if programs are not constructed ahead of time.
Loading optimizer without a train program would not work, which would happen if user set mode something other than `train` before calling `prepare`

File: model.py (hunk 7 above)
Comment (willthefrog) [reply model.py]: see comment below about lazy loading removal.

File: model.py (hunk 8 above)
Comment (qingqing01) [reply model.py]: Now, Loss requires lists of outputs and labels as input.  Not like forward in Model, each input has a key.   There is no key-name for labels in Loss. If use dict, the order is uncertain.

File: model.py (hunk 9 above)
Comment (qingqing01) [reply model.py]: Yeah.  see comments for `train和eval复用Model导致的缺点` in beginning of PR.

File: mnist.py (hunk 11 above)
Comment (qingqing01) [reply mnist.py]: This is for dict inputs.

File: model.py (hunk 10 above)
Comment (qingqing01) [reply model.py]: Done, Thanks!

File: mnist.py (hunk 11 above)
Comment (qingqing01) [reply mnist.py]: Done. Thanks!

File: model.py (hunk 12 above)
Comment (qingqing01) [reply model.py]: @willthefrog @heavengate  Make all programs for train、eval、test in prepare

File: model.py [hunk 13]
```diff
@@ -333,79 +310,76 @@ def _run(self, inputs, labels=None, device='CPU', device_ids=None):
...
+                                 feed=feed,
+                                 fetch_list=fetch_list)
         if self.mode == 'test':
             return out[:num_output]
         else:
             return out[:num_output], out[num_output:]
 
-    def _make_program(self, inputs):
+    def prepare(self):
+        modes = ['train', 'eval', 'test']
```
Comment (willthefrog) [inline model.py]: need checking loss and label availability before build train program
